README.md
Toccata
=======

A Clojure-inspired Lisp dialect that compiles to native executable using the Clang compiler

Check the HISTORY.md file for a detailed description.

This is very much WFM code. Not suitable for any useful purpose.

You can learn more about Toccata by following the [blog here](http://toccata.io)


HISTORY.md
# Rationale

The first (almost) usuable version of Toccata is finally nearing completion. By reading
through the commits and the entries in this file, you should be able to watch Toccata
grow from nothing to a usable language. This is not a historical record. The real
development effort (and resulting code) was very ugly with lots of dead ends, painful bugs,
and tears. (Not really on the tears part, but close)

This repo is to capture the organizational structure of the code.

New entries will be added to the end, and so should be read from the top down.

Happy Reading and Pull Requests welcome!

Jim Duey

# Story

## Entry 1:

In the beginning, God created the Heavens and the Earth. Gen. 1:1

Unlike in Genesis, a lot of work has preceded this commit. Until this compiler is
self-hosting, you won't be able to compile and run it. Nor compile the various regression
tests. I'm using the previous version of Toccata to cross-compile this code until that point.

Now, let us begin.

The `regression-tests` directory will contain a collection of Toccata programs that
exercise various parts of the compiler. `main1.toc` is the simplest program possible.
But already, there's a ton of code required. We have to ...
    open the file
    read it
    parse the contents into a nested list structure
    convert the parsed structure to an AST (Abstract Syntax Tree)
    traverse the AST and emit C code to stdout

If the C code has been written to a file, it can be compiled to an executable using
LLVM or GCC.

So `toccata.toc` is what it takes to do the above steps. This is about as simple a
compiler for Toccata as possible. Everything else builds from there.


## Entry 2:

When an expression is emitted, it is converted to a list of data structures. Each data structure
has the following
    A string that holds the C variable the result of the expression is assigned to
    A list of strings of C code to declare static C values
    A list of strings of C code that produces the expressions result
    A number that signifies the type of the result of the expression
    A hash-map of of C variables that are used in the expression to the number of times their used

Also, `inline` expressions can appear at the top level of a file just like the `main` expression

## Entry 3:

Comment blocks are AST nodes also, this allows easy literate programming and code formatting

## Entry 4:

Defining symbols is the next step. Toccata source files are called modules and eventually there will be
a system for compiling multiple modules into a program. The C variable for each symbol in a module
is added to the global context so it can be looked up later.

## Entry 5:

Now we're able to assign numbers to symbols

## Entry 6:

And now string literals to symbols

## Entry 7:

When doing the next commit, I discovered a need to refactor some code

## Entry 8:

And now we're finally getting somewhere. This is the first point that we can define a function and call it. Looking at this commit, you'll see the complexity start to increase. And we've just scratched the surface. This is where `core.c` makes it's appearance, which will hold all of the C code that is the foundation of Toccata. Watch it explode in successive commits.

## Entry 9:

Since we compile expressions that are calls to functions, we need to handle assiging the results to global static symbols. Which means we need to initialize them at run time before `main` runs. So this commit adds that machinery. It also adds the `core.toc` standard library.

## Entry 10:

Now we'll add a couple of low-level functions that we can call. We'll add the functionality to allocate and free a few primitive value types.

## Entry 11:

Add a fn to compare integers for equality. Which means we need to add the Maybe type as well. And add malloc/free for Functions and Lists.

## Entry 12:

Add `let` expressions to bind values to symbols

## Entry 13:

Inline calls to `list` and `vector` so that those functions don't actually get called to create them. And test their destructuring in let and function parameters.

## Entry 14:

Add `do` expressions for side-effecting code

## Entry 15:

Selecting an expression based on some condition is one of the fundamental features a language must have. So we add `and` & `or` expressions. `or` is the only expression that is lazy. Successive expressions in it's body are only executed after previous expressions evaluate to `nothing`. The first expression that evaluates to a `maybe` value produces the result of the `or` expression. Likewise, `and` is the only expression that short circuits when the first expression returns `nothing`. If all expressions return `maybe` values, the final one becomes the value the `and` expression evaluates to.

## Entry 16:

Though they don't do anything yet, 'assert' expressions are going to play a big role later on. The compiler will use them to do all kinds of static analysis.

## Entry 17:

And now we're finally into the good stuff. Toccata lifts/borrows/steals the idea of `protocols` from Clojure. This commit lets protocols be defined. Almost all of the core library is implemented as protocol functions, which has some very nice properties we'll see later. In Toccata, protocol functions can specify a default impleenation that will be executed if no type-specific protocol implementation is provided.

## Entry 18:

The first hints of where I want to take Toccata appear in `core.toc`.

## Entry 19:

A rather large commit to enable extending types with additional protocol function implementations

## Entry 20:

Well that took longer than I expected. Which is a common occurance when writing a compiler. So many 'little' things lead to lots of other, bigger, things. Everything is connected and a lot has to be done before anything works. In any case, this commit adds variadic functions and fixes list/vector destructuring.

## Entry 21:


core.h

#define CHECK_MEM_LEAK 1

#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <pthread.h>
#include <stdint.h>
#include <inttypes.h>
#include <sys/wait.h>

extern void abort();

#define VECTOR_ARRAY_LEN 32

typedef void (Destructor)(void *);
typedef struct Value {int64_t type; int32_t refs; struct Value* next;} Value;
typedef struct {int64_t type; int32_t refs; int64_t numVal;} Integer;
typedef struct {int64_t type; int32_t refs; int64_t len; Integer *hash; char buffer[0];} String;
typedef struct {int64_t type; int32_t refs; int64_t len; Integer *hash; Value *source; char *buffer;} SubString;
typedef struct List {int64_t type; int32_t refs; int64_t len; Value* head; struct List *tail;} List;
typedef struct {int64_t type; int32_t refs; int count; List *closures; int variadic; void *fn;} FnArity;
typedef struct {int64_t type; int32_t refs; char *name; int64_t arityCount; FnArity *arities[];} Function;
typedef struct {int64_t type; int32_t refs; Value* value;} Maybe;
typedef struct {int64_t type; int32_t refs; Value *array[VECTOR_ARRAY_LEN];} VectorNode;
typedef struct {int64_t type; int32_t refs; int32_t count; int8_t shift; int64_t tailOffset;
                VectorNode *root; Value *tail[VECTOR_ARRAY_LEN];} Vector;
typedef struct {int64_t type; int32_t refs; int32_t bitmap; Value *array[];} BitmapIndexedNode;
typedef struct {int64_t type; int32_t refs; Value *array[32];} ArrayNode;
typedef struct {int64_t type; int32_t refs; int16_t count; Value *array[];} HashCollisionNode;
typedef struct {int64_t type; int32_t refs; Value *result; List *actions;
                pthread_cond_t delivered; pthread_mutex_t access;} Promise;
typedef struct {int64_t type; int32_t refs; Value *action; Value* errorCallback; List *actions;
                Value *result; pthread_cond_t delivered; pthread_mutex_t access;} Future;
typedef struct {int64_t type; int32_t refs;
                Value *val; List* input; List *output;
                pthread_mutex_t access;} Agent;
typedef struct {int64_t type; int32_t refs; Value *typeArgs; int implCount;
                Value* impls[];} ReifiedVal;
typedef struct {int64_t type; int32_t refs; void *ptr; Destructor *destruct;} Opaque;

Integer const0;
Value *const0Ptr;

typedef struct {int64_t type; Value *implFn;} ProtoImpl;
typedef struct {int64_t implCount; ProtoImpl impls[];} ProtoImpls;

typedef Value *(FnType0)(List *);
typedef Value *(FnType1)(List *, Value *);
typedef Value *(FnType2)(List *, Value *, Value *);
typedef Value *(FnType3)(List *, Value *, Value *, Value *);
typedef Value *(FnType4)(List *, Value *, Value *, Value *, Value *);
typedef Value *(FnType5)(List *, Value *, Value *, Value *, Value *, Value *);
typedef Value *(FnType6)(List *, Value *, Value *, Value *, Value *, Value *, Value *);
typedef Value *(FnType7)(List *, Value *, Value *, Value *, Value *, Value *, Value *, Value *);
typedef Value *(FnType8)(List *, Value *, Value *, Value *, Value *, Value *, Value *, Value *, Value *);
typedef Value *(FnType9)(List *, Value *, Value *, Value *, Value *, Value *, Value *, Value *, Value *, Value *);

typedef struct {
  List *tail;
  pthread_mutex_t access;} extractCache;

typedef struct {
  int64_t sym_counter;} intGenerator;

Value *nothing;
Value *maybeNothing;
int32_t refsInit;
int32_t staticRefsInit;
int32_t refsError;

#define UnknownType 0
#define IntegerType 1
#define StringType 2
#define FnArityType 3
#define FunctionType 4
#define SubStringType 5
#define ListType 6
#define MaybeType 7
#define VectorType 8
#define VectorNodeType 9
#define SymbolType 10
#define BitmapIndexedType 11
#define ArrayNodeType 12
#define HashCollisionNodeType 13
#define HashMapType 14
#define PromiseType 15
#define FutureType 16
#define AgentType 17
#define OpaqueType 18
#define TypeCount 19

FILE *outstream;
List *empty_list;
Vector *empty_vect;
BitmapIndexedNode emptyBMI;

struct {List* input; List* output;
        pthread_mutex_t mutex; pthread_cond_t notEmpty;} futuresQueue;
Future shutDown;
int8_t mainThreadDone;

int cleaningUp;

#ifdef CHECK_MEM_LEAK
int64_t malloc_count;
int64_t free_count;
void freeAll();
void freeGlobal(Value *x);
#endif
void dec_and_free(Value *v, int deltaRefs);
Value *incRef(Value *v, int deltaRefs);

Value *proto1Arg(ProtoImpls *protoImpls, char *name, Value *arg0,
                 char *file, int64_t line);
Value *proto2Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1,
                 char *file, int64_t line);
Value *proto3Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 char *file, int64_t line);
Value *proto4Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, char *file, int64_t line);
Value *proto5Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, Value *arg4, char *file, int64_t line);
Value *proto6Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, Value *arg4, Value *arg5, char *file, int64_t line);
Value *proto7Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, Value *arg4, Value *arg5, Value *arg6,
                 char *file, int64_t line);
Value *proto8Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, Value *arg4, Value *arg5, Value *arg6, Value *arg7,
                 char *file, int64_t line);
Value *proto9Arg(ProtoImpls *protoImpls, char *name, Value *arg0, Value *arg1, Value *arg2,
                 Value *arg3, Value *arg4, Value *arg5, Value *arg6, Value *arg7,
                 Value *arg8, char *file, int64_t line);

void prefs(char *tag, Value *v);

Value *(*equalSTAR)(List *closures, Value *, Value *);
Value *(*assoc)(List *closures, Value *, Value *, Value *, Value *, Value *);
Value *(*dissoc)(List *closures, Value *, Value *, Value *, Value *);
Value *(*get)(List *, Value *, Value *, Value *, Value *, Value *);
Value *(*sha1)(List *, Value *);
Value *(*hashSeq)(List *, Value* n, Value *s);
Value *(*count)(List *, Value* n);
// Value *(*apply)(List *closures, Value *f, Value *args);
Value *(*invoke0Args)(List *closures, Value *f);
Value *(*invoke1Arg)(List *closures, Value *f, Value* arg);
Value *(*invoke2Args)(List *closures, Value *f, Value* arg0, Value* arg1);
Value *(*type_name)(List *closures, Value *t);
Value *(*fn_apply)(List *closures, Value *f, Value *args);

Value *my_malloc(int64_t sz);
List *malloc_list();
Value *vectSeq(Vector *vect, int index);
FnArity *malloc_fnArity();
Function *malloc_function(int arityCount);
String *malloc_string(int len);
FnArity *findFnArity(Value *fnVal, int64_t argCount);
ReifiedVal *malloc_reified(int implCount);
Promise *malloc_promise();

void startWorkers();
void replaceWorker();
void waitForWorkers();
char *extractStr(Value *v);
Value *isInstance(Value *arg0, Value *arg1);
Value *intValue(int64_t n);
Value *prSTAR(Value *);
Value *prErrSTAR(Value *);
Value *add_ints(Value *arg0, Value *arg1);
Value *integer_str(Value *arg0);
Value *integer_EQ(Value *arg0, Value *arg1);
Value *integer_LT(Value *arg0, Value *arg1);
Value *integerValue(int64_t n);
Vector *vectConj(Vector *vect, Value *val);
Vector *mutateVectConj(Vector *vect, Value *val);
Value *vectStore(Vector *vect, unsigned index, Value *val);
Value *vectorReverse(Value *arg0);
List *listCons(Value *x, List *l);
void destructValue(char *fileName, char *lineNum, Value *val, int numArgs, Value **args[]);
Value *maybe(List *closures, Value *arg0, Value *arg1);
int8_t isNothing(Value *v);
Value *strCount(Value *arg0);
Value *strEQ(Value *arg0, Value *arg1);
Value *strList(Value *arg0);
Value *strVect(Value *arg0);
Value *checkInstance(int64_t typeNum, Value *arg1);
Value *listMap(Value *arg0, Value *arg1);
Value *listConcat(Value *arg0);
Value *car(Value *arg0);
Value *cdr(Value *arg0);
Value *integerLT(Value *arg0, Value *arg1);
Value *integerSha1(Value *arg0);
Value *bitAnd(Value *arg0, Value *arg1);
Value *bitOr(Value *arg0, Value *arg1);
Value *addIntegers(Value *arg0, Value *arg1);
Value *listEQ(Value *arg0, Value *arg1);
int8_t equal(Value *v1, Value *v2);
Value *maybeExtract(Value *arg0);
Value *fnApply(Value *arg0, Value *arg1);
Value *maybeApply(Value *arg0, Value *arg1);
Value *maybeEQ(Value *arg0, Value *arg1);
Value *maybeMap(Value *arg0, Value *arg1);
Value *strSha1(Value *arg0);
Value *escapeChars(Value *arg0);
Value *subs2(Value *arg0, Value *arg1);
Value *subs3(Value *arg0, Value *arg1, Value *arg2);
Value *strSeq(Value *arg0);
Value *strReduce(Value *s0, Value *x1, Value *f2);
Value *strVec(Value *arg0);
Value *strLT(Value *arg0, Value *arg1);
Value *vectorGet(Value *arg0, Value *arg1);
Value *symbol(Value *arg0);
Value *symbolSha1(Value *arg0);
Value *symEQ(Value *arg0, Value *arg1);
Value *symLT(Value *arg0, Value *arg1);
Value *stringValue(char *s);
Value *opaqueValue(void *ptr, Destructor *destruct);
Value *maybeInvoke(Value *arg0, Value *arg1, Value *arg2);
Value *listFilter(Value *arg0, Value *arg1);
List *reverseList(List *input);
Value *bmiHashSeq(Value *arg0, Value *arg1);
Value *bmiCount(Value *arg0);
Value *bmiAssoc(Value *arg0, Value *arg1, Value *arg2, Value *arg3, Value* arg4);
Value *bmiGet(Value *arg0, Value *arg1, Value *arg2, Value *arg3, Value *arg4);
Value *bmiDissoc(Value *arg0, Value* arg1, Value* arg2, Value* arg3);
Value *arrayNodeAssoc(Value *arg0, Value *arg1, Value *arg2, Value* arg3, Value *arg4);
Value *collisionAssoc(Value *arg0, Value *arg1, Value *arg2, Value *arg3, Value *arg4);
Value *hashMapGet(Value *arg0, Value *arg1);
Value *arrayNodeGet(Value *arg0, Value *arg1, Value *arg2, Value *arg3, Value *arg4);
Value *collisionCount(Value *arg0);
Value *arrayNodeCount(Value *arg0);
Value *collisionSeq(Value *arg0, Value *arg1);
Value *collisionDissoc(Value *arg0, Value *arg1, Value *arg2, Value *arg3);
Value *collisionGet(Value *arg0, Value *arg1, Value *arg2, Value *arg3, Value *arg4);
Value *arrayNodeSeq(Value *arg0, Value *arg1);
Value *arrayNodeDissoc(Value *arg0, Value *arg1, Value *arg2, Value *arg3);
Value *deliverPromise(Value *arg0, Value *arg1);
Value *extractPromise(Value *arg0);
Value *promiseDelivered(Value *arg0);
Value *extractFuture(Value *arg0);
Value *makeFuture(Value *arg0);
Value *makeAgent(Value *arg0);
Value *extractAgent(Value *arg0);
void scheduleAgent(Agent *agent, List *action);
void freeExtractCache(void *cachePtr);
void freeIntGenerator(void *ptr);
Value *addPromiseAction(Promise *promise, Value *action);
Value *dynamicCall1Arg(Value *f, Value *arg);
Value *deliverFuture(Value *fut, Value *val);
Value *addFutureAction(Future *p, Value *action);
String *nullTerm(Value *s);
